src/combinatorics/simple_graph/acyclic.lean

```diff
@@ -0,0 +1,137 @@
+/-
+Copyright (c) 2022 Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kyle Miller
+-/
+import combinatorics.simple_graph.connectivity
+/-!
+
+# Acyclic graphs and trees
+
+This module introduces *acyclic graphs* (a.k.a. *forests*) and *trees*.
+
+The structure of the proofs for `simple_graph.is_acyclic` and `simple_graph.is_tree`, including
+supporting lemmas about `simple_graph.is_bridge`, generally follows the description for these
+theorems for multigraphs from [Chou1994].
+
+## Main definitions
+
+* `simple_graph.is_acyclic` is a predicate for a graph having no cyclic walks
+* `simple_graph.is_tree` is a predicate for a graph being a tree (a connected acyclic graph)
+
+## Main statements
+
+* `simple_graph.is_acyclic_iff` characterizes acyclicity in terms of uniqueness of paths between
+  pairs of vertices.
+* `simple_graph.is_acyclic_iff_forall_edge_is_bridge` characterizes acyclicity in terms of every
+  edge being a bridge edge.
+* `simple_graph.is_tree_iff` characterizes trees in terms of existence and uniqueness of paths
+  between pairs of vertices from a nonempty vertex type.
+
+## Tags
+
+acyclic graphs, trees
+-/
+
+universes u v
+
+namespace simple_graph
+variables {V : Type u} (G : simple_graph V)
+
+/-- A graph is *acyclic* (or a *forest*) if it has no cycles. -/
+def is_acyclic : Prop := ∀ (v : V) (c : G.walk v v), ¬c.is_cycle
+
+/-- A *tree* is a connected acyclic graph. -/
+def is_tree : Prop := G.connected ∧ G.is_acyclic
+
+variables {G}
+
+lemma is_acyclic_iff_forall_adj_is_bridge :
+  G.is_acyclic ↔ ∀ {v w : V}, G.adj v w → G.is_bridge ⟦(v, w)⟧ :=
+begin
+  simp_rw [is_bridge_iff_adj_and_forall_cycle_not_mem],
+  split,
+  { intros ha v w hvw,
+    apply and.intro hvw,
+    intros u p hp,
+    exact absurd hp (ha _ p), },
+  { rintros hb v (_ | @⟨_, _, _, ha, p⟩) hp,
+    { exact hp.not_of_nil },
+    { specialize hb ha,
+      apply hb.2 _ hp,
+      rw [walk.edges_cons],
+      apply list.mem_cons_self } },
+end
+
+lemma is_acyclic_iff_forall_edge_is_bridge :
+  G.is_acyclic ↔ ∀ (e ∈ G.edge_set), G.is_bridge e :=
```

Comment:
```suggestion
  G.is_acyclic ↔ ∀ e ∈ G.edge_set, G.is_bridge e :=
```
or, even better,
```suggestion
  G.is_acyclic ↔ ∀ ⦃e⦄, e ∈ G.edge_set → G.is_bridge e :=
```

Reply:
I personally like to write bounded quantifiers like `∀ (e ∈ G.edge_set)` and I believe this is a matter of taste rather than mathlib style.

Regarding weak implicits, I've found I run into odd issues whenever I do try to use them. Do you think this is worth changing this? I'm not sure what's *really* gained other than saving two characters here and there.

Reply:
It is genuinely quite nicer once you try to manipulate the expressions unapplied. See my other comment for why it really is no cost having weak implicits.